### ISSUE TITLE
Problem regarding true data statics computation for FVD

### ISSUE THREAD

**Issue by ParthaEth** [universome/stylegan-v#37]
I ran into trouble trying to reproduce your reported FVD in sky time-lapse dataset. I'll list my problems in the following list

1. If we take the official dataset from - https://github.com/weixiong-ur/mdgan, the frames are of size 360X640. As far as I can tell reading the code of your data loader, you take a center crop of 360X360 and then resize it down to 256X256. Is that correct?

2. While evaluating FVD which set do you use to compute real statistics?
    1. In the paper in appendix C you write you choose a *single* clip per video, if we do that with clip length of 128 frames we get about 1K clips from the train set. Bur for FVD we obviously need at least 2048 videos. Therefore, closest we can get is to use maximum 4 clips per video, and we then get 2069 clips. Let's call this is FVD_comp_set.
    2. Should we use this set to compute FVD_16f as well? If so do you take first 16 consecutive frames per clip?
    3. If we compute FVD against this set (FVD_comp_set) of the given [checkpoint](https://disk.yandex.ru/d/v7MS7zu4mmZxXw) we get fvd_2048_16f: 49.94 and fvd_2048_128f:108.82, whereas the mentioned FVD in the paper for the same metrics are 79.52 and 197.0 respectively.
    4. Next, I downloaded the sky time-lapse dataset you provided in the [checkpoint link](https://disk.yandex.ru/d/v7MS7zu4mmZxXw). Then reloaded the network checkpoint and attempted to fine-tune it. This evaluates the metric and that reports fvd_2048_16f to be 46.46. Therefore, random train samples as true stats computation of FVD seems not to be the case either.
   
My request is can you detail the way you computed FVD fro the paper so I can compare our method to yours?

**Comment by universome** [comment]
Hi Partha, thanks for reaching out!

1. Yes, that's correct.
2. I guess our metric naming led to some misunderstanding. We compute two metrics:
```python
@register_metric
def fvd2048_16f(opts):
    opts.dataset_kwargs.update(max_size=None, xflip=False)
    fvd = frechet_video_distance.compute_fvd(opts, max_real=2048, num_gen=2048, num_frames=16)
    return dict(fvd2048_16f=fvd)

@register_metric
def fvd2048_128f(opts):
    opts.dataset_kwargs.update(max_size=None, xflip=False)
    fvd = frechet_video_distance.compute_fvd(opts, max_real=2048, num_gen=2048, num_frames=128)
    return dict(fvd2048_128f=fvd)
```
The first one, `fvd2048_16f`, uses at most 2048 real clips (`max_real=2048`) and 2048 fake clips. If there are fewer real clips available for the given amount of frames (for `fvd2048_16f` — 16 ones), we discard the entire videos by setting [`discard_short_videos=True`](https://github.com/universome/stylegan-v/blob/3fecd69c602e1cda204357201461c0fb0a634909/src/metrics/frechet_video_distance.py#L28). Exactly the same logic applies to `fvd2048_128f`, which in the end uses fewer clips for SkyTimelapse. The motivation for this is that some datasets are heavily dominated by very-long videos which would heavily bias the statistics (e.g., SkyTimelapse contains a couple of several-hours-long videos, if I remember correctly). To answer your questions more specifically:
    1. We use a single clip per video, we do not use 2048 videos for `fvd2048_128f` for SkyTimelapse. The number `2048` in the metric name `fvd2048_128f` denotes how many fake videos we use. The notation is borrowed from StyleGAN-s which call their metrics something like "fid50k_full". I guess, a better metric notation in our case would be "fvd2048_max2048_128f". We used `max_real=2048` with the only purpose of being closer to the prior work — I believe in general it would be best to use all the available videos, like the community does for FID.
    2. `fvd2048_16f` would likely use a broader set of videos since fewer videos will be filtered away by length.
    3. To be honest, it's hard to tell why there is such a difference. Maybe your `FVD_comp_set` has more robust statistics since it was computed on more clips and hence is more generous.
    4. The checkpoint we provided could be a bit better or worse compared to the one that was used during the paper submission.

I would suggest to either just generate videos from our provided checkpoint and compute fake statistics on top of them or use the numbers from the paper — both options seem fine. Frankly, right now it feels like the best way to compute FVD is by using all the available real videos, but using the same amount of clips per video so as not to bias the statistics towards the long videos (which are likely outliers). We didn't use this way for "legacy" reasons (i.e. to keep closer to the prior work).

**Comment by ParthaEth** [comment]
Thanks for the detailed answer. I'll report FVD in the same way you did for compatibility. In sup-mat perhaps we report FVD against a set that has at least 2048 real samples. I remember FID was very unstable when in early days people computed using 10K or even less samples. I suppose FVD has similar behavior.

**Comment by ParthaEth** [comment]
I am still confused about 

 I downloaded the sky time-lapse dataset you provided in the [checkpoint link](https://disk.yandex.ru/d/v7MS7zu4mmZxXw). Then reloaded the network checkpoint and attempted to fine-tune it. This evaluates the metric and that reports fvd_2048_16f to be *46.46*. Therefore, random train samples as true stats computation of FVD seems not to be the case either.

How can this be so low? Any suggestions?

**Comment by universome** [comment]
Most likely the final provided checkpoint was trained with better hyperparameters than the one used for the paper.
